doc/releases/changelog-dev.md

```diff
@@ -106,21 +107,52 @@
     significant performance improvement when executing the QNode on remote
     quantum hardware.
 
+  - When decomposing the circuit, the default decomposition strategy will prioritize
+    decompositions that result in the smallest number of parametrized operations
+    required to satisfy the differentiation method. Additional decompositions required
+    to satisfy the native gate set of the quantum device will be performed later, by the
+    device at execution time. While this may lead to a slight increase in classical processing,
+    it significantly reduces the number of circuit evaluations needed to compute
+    gradients of complex unitaries.
+
   In an upcoming release, this QNode will replace the existing one. If you come across any bugs
   while using this QNode, please let us know via a [bug
   report](https://github.com/PennyLaneAI/pennylane/issues/new?assignees=&labels=bug+%3Abug%3A&template=bug_report.yml&title=%5BBUG%5D)
   on our GitHub bug tracker.
 
   Currently, this beta QNode does not support the following features:
 
-  - Circuit decompositions
   - Non-mutability via the `mutable` keyword argument
   - Viewing specifications with `qml.specs`
   - The `reversible` QNode differentiation method
   - The ability to specify a `dtype` when using PyTorch and TensorFlow.
 
   It is also not tested with the `qml.qnn` module.
 
+* Two new methods were added to the Device API, allowing PennyLane devices
+  increased control over circuit decompositions.
+  [(#1651)](https://github.com/PennyLaneAI/pennylane/pull/1651)
+
+  - `Device.expand_fn(tape) -> tape`: expands a tape such that it is supported by the device. By
+    default, performs the standard device-specific gate set decomposition done in the default
+    QNode. Devices may overwrite this method in order to define their own decomposition logic.
```

Comment:
Oh, this is a good point 🤔 

At the moment yes, but it looks a bit hacky. You could do something like this:

```python
>>> def my_custom_expand_fn(tape, **kwargs):
...     print("hello")
...     return tape
>>> qnode.device.expand_fn = my_custom_expand_fn
>>> qnode(0.5)
hello
tensor(0.87758256, requires_grad=True)
```

Hmmm 🤔 Do you think this will be useful?

Reply:
@glassnotes I think @anthayes92 might be on to something re: custom decompositions....

Reply:
I mean, we could even support something like how you currently ['register' QNode execution wrappers](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.batch_transform.html#pennylane.batch_transform.custom_qnode_wrapper) while writing a batch transform

```python
dev = qml.device("default.qubit", wires=2)

@dev.custom_expand
def my_expansion_function(tape, **kwargs):
    ...
    return tape

# from now on, the custom expansion is called whenever
# the device is executed.
```

This is more powerful (too powerful?) compared to a dictionary of gates to custom decompositions. But I still have some question marks:

1. Should this _replace_ the device expansion?
2. If it doesn't, does it come before the device expansion? This way unsupported gates are finally decomposed down to device native gates. Or should it come _after_ the device expansion? Execution would then fail if the custom decomp results in an unsupported gate.
3. Rather than changing the device, does it make more sense to pass a custom decomposition to the QNode? E.g.,

   ```python
   @qml.qnode(dev, expansion_strategy=my_custom_expansion)

   # or

   @existing_qnode.register_expansion
   def my_custom_expansion(...):
   ```

Reply:
> This is more powerful (too powerful?) compared to a dictionary of gates to custom decompositions.

How would custom decompositions be specified in these cases?

What if we did something like this, which combines a few of the ideas floating around:

```python

custom_decomps = {qml.Hadamard : h_func, qml.CNOT : cnot_func}

def custom_expand(tape, custom_decomps):
    # applies custom decompositions 

qnode.device.set_expand_fn(custom_expand)
```

but where the `device` itself does some sort of internal validation of the decompositions? 

```python
def set_expand_fn(custom_decomps):
    for op, decomp in custom_decomps.items():
       # Ensure all the operations in the decomposition are valid for the device
       ...
       # Register the new decompositions to the operators
       if decomp_is_valid:
           op.register_new_decomposition(decomp)
```

If we do this kind of validation, it ensures that we can apply the expansion *after* the gradient tapes have already been constructed, but with the guarantee that they'll still run on the device.

Reply:
I really like @glassnotes suggestion of having expansion after the gradient tapes have already been constructed! So would the logic here look like: if custom gates are unsupported then decompose to device native gates, so that this is where the guarantee they'll still run on the device comes from?